test/app-luatest/pcall_test.lua

```diff
@@ -0,0 +1,78 @@
+local server = require('luatest.server')
+local t = require('luatest')
+
+local g = t.group()
+
+g.before_all(function()
+    g.server = server:new({alias = 'pcall'})
+    g.server:start()
+end)
+
+g.after_all(function()
+    g.server:stop()
+end)
+
+-- #267: Bad exception catching.
+g.test_pcall_inside_xpcall = function()
+    g.server:exec(function()
+        local function pcalltest()
+            local ERRMSG = "module 'some_invalid_module' not found"
+            local status, msg = pcall(require, 'some_invalid_module')
+            if status == false and msg ~= nil and msg:match(ERRMSG) ~= nil then
+                return 'pcall is ok'
+            else
+                return 'pcall is broken'
+            end
+        end
+
+        local status, msg = xpcall(pcalltest, function(msg)
+            print('error handler:', msg)
+        end)
+        t.assert_equals(status, true, 'pcall inside xpcall')
+        t.assert_equals(msg, 'pcall is ok', 'pcall inside xpcall')
+    end)
+end
+
+g.test_pcall_with_lua_error = function()
+    g.server:exec(function()
+        local status, msg = pcall(function() error('some message') end)
+        t.assert_equals(status, false, 'pcall with Lua error()')
+        t.assert_equals(msg:match('some message'), 'some message',
+                        'pcall with Lua error()')
+    end)
+end
+
+g.test_pcall_with_box_error = function()
+    g.server:exec(function()
+        local ffi = require('ffi')
+
+        local status, msg = pcall(function()
+            box.error(box.error.ILLEGAL_PARAMS, 'some message')
+        end)
+
+        t.assert_equals(status, false, 'status')
+        t.assert_equals(tostring(ffi.typeof(msg)),
+                        'ctype<const struct error &>', 'typeof')
+        t.assert_equals(msg.type, 'IllegalParams', '.type')
+        t.assert_equals(msg.message, 'some message', '.message')
+
+        -- Tarantool 1.6 backward compatibility.
+        t.assert_equals(msg:match('some'), 'some', '.match()')
```

Comment:
I would just drop it. I'm in doubt that we still need Tarantool 1.6 backward compatibility.
Similar assertion is just above.

Reply:
I'm OK with that. Ping @Totktonada, as a guru for backward compatibility.

Reply:
If we're speaking about the API itself (not about the test case), I would say the following.

We can decide to break an API, when we're releasing a new major version. However, we can't do it in a middle of the 3.X series.

The idea here is the following. This 1.6 API is part of 3.2 API, because we support it in 3.2. We can't break 3.2 API in 3.3.

The same from the user perspective: code that work on 3.2 should work on 3.3 (even if it use deprecated APIs).

----

If we're speaking about the test case, my opinion is that we should keep a test while we're support the functionality.

I think that it doesn't worth a discussion around compatibility rules: it is easier to just keep the test case.

Reply:
Ok, thanks for a wordy explanation.
I'll remove comment `-- Tarantool 1.6 backward compatibility.`, because it is confusing.

```diff
--- a/test/app-luatest/pcall_test.lua
+++ b/test/app-luatest/pcall_test.lua
@@ -56,8 +56,6 @@ g.test_pcall_with_box_error = function()
         t.assert_equals(msg.type, 'IllegalParams', 'pcall with box error, type')
         t.assert_equals(msg.message, 'some message',
                         'pcall with box error, message')
-
-        -- Tarantool 1.6 backward compatibility.
         t.assert_equals(msg:match('some'), 'some',
                         'pcall with box error, message')
     end)
```